doc/BluetoothLEStack.md

```diff
@@ -0,0 +1,171 @@
+
+  * Connectable directed
+
+  * Non-connectable undirected
+
+  * Scannable undirected
+
+The driver is _not_ responsible for validating that the payload for these
+advertising types follows any particular specification. Advertising event types
+that require particular interactions at the link-layer with peer devices (e.g.
+scanning or establishing connections) are not permissible:
+
+  * Scan request
+
+  * Scan response
+
+  * Connect request
+
+Scan response are sent automatically if a scan response payload is configured.
+Scan rerequest and connection requests are handled by other parts of the sytsem
+call interface.
+
+To set up an advertisement:
+
+  1. Configure the advertisement payload, type, interval, tx power and,
+     optionally, scan response payload.
+
+     * Advertisement payload `allow`
+
+     * Advertisement type `command`
+
+     * If the advertising type is scannable, you SHOULD configure a scan
+       response payload using `allow`
+
+     * Advertisement interval `command`
+
+     * Advertisement tx power `command`
+
+  2. Start periodic advertising using a `command`
+
+Any changes to the configuration while periodic advertising is happening will
+take effect in a future advertising event. The kernel will use best effort to
+reconfigure advertising in as few events as possible.
+
+To stop advertising
+
+  1. Stop periodic advertising a `command`
+
+### Scanning
+
+### Connection-oriented communication
+
+## Hardware Interface Layer (HIL)
+
+The Bluetooth Low Energy Radio HIL defines a cross-platform interface for
+interacting with on-chip BLE radios (i.e. it does not necessarily work for
+radios on a dedicated IC connected over a bus).
+
+The goal of this interface is to expose low-level details of the radio that are
+common across platforms, except in cases where abstraction is needed for common
+cases to meet timing constraints.
+
+
+```rust
+pub trait BleRadio {
+    /// Sets the channel on which to transmit or recieve packets.
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn set_channel(&self, channel: RadioChannel) -> ReturnCode;
+
+    /// Sets the transmit power
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn set_tx_power(&self, power: u8) -> ReturnCode;
+
+    /// Transmits a packet over the radio
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn transmit_packet(
+        &self,
+        buf: &'static mut [u8],
+        disable: bool,
+        len: usize) -> ReturnCode;
+
+    /// Recieves a packet of at most `buf.len()` size
+    ///
+    /// Returns ReturnCode::EBUSY if the radio is currently transmiting or
+    /// recieving, otherwise ReturnCode::Success.
+    fn receive_packet(&self, buf: &'static mut [u8]) -> ReturnCode;
```

Comment:
This is good with `buf` here but it requires changes to userspace if we want to advertise and scan `concurrently`. Also with this I think we can remove the `static buffers` in `chips` then

Reply:
Also how do you propose to the replace `TakeCell` in the `BLE Capsule`?